File: backend/app/api/v1/tasks.py
import uuid
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_session
from app.core.deps import PaginationParams, get_current_user
from app.crud.task import (
    create_task,
    get_task_public,
    get_task_summary,
    list_tasks,
    soft_delete_task,
    update_task,
)
from app.models.user import User
from app.schemas.paginated import Paginated
from app.schemas.task import TaskCreate, TaskPublicRead, TaskSummary, TaskUpdate

router = APIRouter(prefix="/tasks", tags=["tasks"])


@router.post("/", response_model=TaskPublicRead, status_code=status.HTTP_201_CREATED)
async def create(
    data: TaskCreate,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    return await create_task(db, current_user.id, data)


@router.get("/summary", response_model=TaskSummary)
async def summary(
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    return await get_task_summary(db, current_user.id)


@router.get("/", response_model=Paginated[TaskPublicRead])
async def list_all(
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
    pagination: Annotated[PaginationParams, Depends(PaginationParams)],
    status: str | None = None,
    priority: str | None = None,
    person_id: uuid.UUID | None = None,
    asset_id: uuid.UUID | None = None,
    subscription_id: uuid.UUID | None = None,
    event_id: uuid.UUID | None = None,
    due_before: date | None = None,
    due_after: date | None = None,
    overdue: bool | None = None,
):
    items, total = await list_tasks(
        db,
        current_user.id,
        skip=pagination.skip,
        limit=pagination.limit,
        status=status,
        priority=priority,
        person_id=person_id,
        asset_id=asset_id,
        subscription_id=subscription_id,
        event_id=event_id,
        due_before=due_before,
        due_after=due_after,
        overdue=overdue,
    )
    return Paginated(
        items=items, total=total, skip=pagination.skip, limit=pagination.limit
    )


@router.get("/{task_id}", response_model=TaskPublicRead)
async def get_one(
    task_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    task = await get_task_public(db, task_id, current_user.id)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.patch("/{task_id}", response_model=TaskPublicRead)
async def patch(
    task_id: uuid.UUID,
    data: TaskUpdate,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    task = await update_task(db, task_id, current_user.id, data)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete(
    task_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    task = await soft_delete_task(db, task_id, current_user.id)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

File: frontend/src/api/tasks.ts
import { api } from "./client";
import type { Paginated, TermSlim } from "./types";

export interface TaskPublicRead {
  id: string;
  owner_id: string;
  title: string;
  description: string | null;
  status: string;
  priority: string;
  due_date: string | null;
  completed_at: string | null;
  person_id: string | null;
  asset_id: string | null;
  subscription_id: string | null;
  tags: TermSlim[];
  created_at: string;
  updated_at: string;
}

export interface TaskSummary {
  total: number;
  by_status: Record<string, number>;
  overdue: number;
  due_today: number;
}

export interface TaskCreatePayload {
  title: string;
  description?: string | null;
  status?: string;
  priority?: string;
  due_date?: string | null;
  person_id?: string | null;
  asset_id?: string | null;
  subscription_id?: string | null;
  tags?: string[];
}

export type TaskUpdatePayload = Partial<TaskCreatePayload>;

export function listTasks(params: {
  skip?: number;
  limit?: number;
  status?: string;
  priority?: string;
  person_id?: string;
  asset_id?: string;
  subscription_id?: string;
} = {}): Promise<Paginated<TaskPublicRead>> {
  const q = new URLSearchParams();
  if (params.skip !== undefined) q.set("skip", String(params.skip));
  if (params.limit !== undefined) q.set("limit", String(params.limit));
  if (params.status) q.set("status", params.status);
  if (params.priority) q.set("priority", params.priority);
  if (params.person_id) q.set("person_id", params.person_id);
  if (params.asset_id) q.set("asset_id", params.asset_id);
  if (params.subscription_id) q.set("subscription_id", params.subscription_id);
  const qs = q.toString();
  return api.get<Paginated<TaskPublicRead>>(`/tasks${qs ? `?${qs}` : ""}`);
}

export function getTaskSummary(): Promise<TaskSummary> {
  return api.get<TaskSummary>("/tasks/summary");
}

export function getTask(taskId: string): Promise<TaskPublicRead> {
  return api.get<TaskPublicRead>(`/tasks/${taskId}`);
}

export function createTask(data: TaskCreatePayload): Promise<TaskPublicRead> {
  return api.post<TaskPublicRead>("/tasks", data);
}

export function updateTask(taskId: string, data: TaskUpdatePayload): Promise<TaskPublicRead> {
  return api.patch<TaskPublicRead>(`/tasks/${taskId}`, data);
}

export function deleteTask(taskId: string): Promise<void> {
  return api.delete(`/tasks/${taskId}`);
}

